### PR TITLE
Refactor Emval implementation to avoid JS heap allocations.

### PR DESCRIPTION
**Pull request by mrolig5267319**
Refactors the JS implementation of emscripten::val to use a single array to hold both the JavaScript values and the refcount, rather than using the array from a HandleAllocator with JS objects holding pairs of value and refcount.

Motivation is to reduce overhead for allocating such objects and GC pressure from short-lived objects. This design continues to use a single array that never shrinks to hold the JS values an refcounts, and a freelist as a stack of available handles to reuse array entries. Once in a steady state, there should be approximately no allocations required to add and remove entries.

This changes Emval to not reuse the HandleAllocator, so some logic from that class is duplicated, with revision for the array to hold two entries per handle.

In my target application, this was the only use of HandleAllocator, and I saw a decrease in JS code size as a result (both before and after closure compilation). Even when I left in the dependency on HandleAllocator to keep that code in the binary to approximate impact of projects that still use HandleAllocator, I observe a smaller code size with this change.

I do not have a benchmark to demonstrate the improvement, but believe on the surface avoiding heap allocations is obviously better, especially if there isn't code size growth as a result.

/cc @brendandahl @sbc100

### CODE REVIEW

File: src/embind/emval.js
Comment (sbc100) [inline src/embind/emval.js:94]: Should we just remove this line and initialize all the the reserved handles to refcount == 1? 

That will safe on code size and remove an extra branch here.. at the cost of redundantly storing/tracking the refcount of the reserved types (could could assert that the refcount of a reserved type never hits zero too).

File: src/embind/emval.js
Comment (sbc100) [inline src/embind/emval.js:25]: Rather than adding this here (which will make it exist at runtime).. you do this at the top level:

```
{{{ 
globalThis.EMVAL_RESERVED_HANDLES = 5;
null;
}}}
```

And then reference the value with ```{{{ EMVAL_RESERVED_HANDLES }}}```

**Comment by sbc100** [comment]
Looks like you will need to run `test_minimal_runtime_code_size_hello_embind_val` with `--rebase`.... perhaps some of the suggestions I made will help minimize any delta there.

**Comment by RReverser** [comment]
Hm it would be good to see some benchmarks before/after this change. JS engines generally optimise frequent creation / deletion of small objects very well (because it's a very common usage), and they have extra optimisations for arrays of monomorphic shape, so I'd expect GC pressure not to come into play here at all, unlike in interpreted languages.

Given that this somewhat lowers readability of the code and makes array non-monomorphic, I think some simple benchmarks showing a clear performance win would be warranted before merging.

**Comment by RReverser** [comment]
This also reminds me that I should find time (if only) and finish https://github.com/emscripten-core/emscripten/pull/20447 to move reference counting to C++ entirely...

**Review by sbc100 — COMMENTED** [review]
lgtm assuming this actually leads to performance improvement.   Are you able do some benchmarking?

File: src/embind/emval.js
Comment (mrolig5267319) [reply src/embind/emval.js:75]: Added a comment, this is because the handle 1 is for the value `undefined`.

File: src/embind/emval.js
Comment (mrolig5267319) [reply src/embind/emval.js]: Was also on my mind, but was trying to minimize the changes. Seems better to minimize the code paths though.

**Comment by mrolig5267319** [comment]
@RReverser Thanks for the reference to your other [draft pull request](https://github.com/emscripten-core/emscripten/pull/20447) with a benchmark.

I've run that benchmark and I observe a ~20% improvement to Allocation cost:

Before:

```
Allocation cost: 92 ns
Copying cost: 61 ns
```

After:

```
Allocation cost: 76 ns
Copying cost: 59 ns
```

**Comment by mrolig5267319** [comment]
Also, @sbc100 , the failures in `test_minimal_runtime_code_size_hello_embind_val` were reflecting the expected *decrease* in code size, not a regression.

**Comment by mrolig5267319** [comment]
Fixed test_embind_val ... the reason we special case the increment/decrement for reserved values is that C++ will use them without incrementing the refcount.  Had to add that back to the JS.

**Comment by mrolig5267319** [comment]
@sbc100 @RReverser -- looks like all tests are passing, anything else I should do to move this forward?

File: src/embind/emval.js
Comment (juj) [inline src/embind/emval.js:44]: I think it should be possible to remove all the `/ 2` and `* 2` business for further slightly smaller code size and micro-perf win. The fact that the returned handles would "only" have a 1 billion range instead of 2 billion range is likely not ever possible to exhaust - that 1 billion range would already mean more than 16 gigabytes of memory usage in the emval_handles array. (2*double for each handle).

Works either way though.

**Review by juj — APPROVED** [review]
LGTM

File: src/embind/emval.js
Comment (sbc100) [reply src/embind/emval.js:44]: Oh I see, you mean handing back only even numbers as a handles.   Nice idea!

(Then I guess we can assert that handles are even when we receive them from the user).

File: src/embind/emval.js
Comment (sbc100) [inline src/embind/emval.js:21]: Can we keep this at 5 to avoid confusion?

It only two places it is used seem like they would be fine with that.

Comment (mrolig5267319) [reply src/embind/emval.js]: Done.

File: src/embind/emval.js
Comment (mrolig5267319) [reply src/embind/emval.js]: Updated as suggested, did require updates to the .h file for the new even handles for reserved values.

**Pull request merged**
